tests/integration/extenders/ServiceProviderTest.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Tests\integration\extenders;
+
+use Flarum\Foundation\AbstractServiceProvider;
+use Flarum\Tests\integration\TestCase;
+use Flarum\Extend;
+use Illuminate\Contracts\Container\BindingResolutionException;
+
+class ServiceProviderTest extends TestCase
+{
+    /**
+     * @test
+     */
+    public function provider_doesnt_work_by_default()
```

Comment:
I think the important thing to check in these tests is the order in which core service providers, service provider register / extenders, and service provider boot resolve. Perhaps:

In one service provider, have `register` override some (existing) singleton from core, and check that that is overriden
In another, have `boot` override it again and check that what `boot` does is the final result